syntax = "proto3";

package game;

// 서비스 정의
service GameService {
  // 유저 정보 조회
  rpc GetUser (UserRequest) returns (UserResponse) {}
  // 새 유저 추가
  rpc AddUser (AddUserRequest) returns (AddUserResponse) {}
  // 점수 업데이트
  rpc UpdateScore (UpdateScoreRequest) returns (UpdateScoreResponse) {}
  // 랭킹 조회
  rpc GetRanking (RankingRequest) returns (RankingResponse) {}
}

// 유저 정보 요청
message UserRequest {
  int32 user_id = 1;
}

// 유저 정보 응답
message UserResponse {
  int32 user_id = 1;
  string username = 2;
  int32 score = 3;
  string created_at = 4;
}

// 새 유저 추가 요청
message AddUserRequest {
  string username = 1;
}

// 유저 추가 응답
message AddUserResponse {
  bool success = 1;
  string message = 2;
  int32 user_id = 3;
}

// 점수 업데이트 요청
message UpdateScoreRequest {
  int32 user_id = 1;
  int32 score = 2;
}

// 점수 업데이트 응답
message UpdateScoreResponse {
  bool success = 1;
  string message = 2;
  int32 new_score = 3;
}

// 랭킹 요청
message RankingRequest {
  int32 limit = 1; // 상위 몇 명을 조회할지
}

// 랭킹 응답
message RankingResponse {
  message RankingEntry {
    int32 rank = 1;
    int32 user_id = 2;
    string username = 3;
    int32 score = 4;
  }
  repeated RankingEntry rankings = 1;
}